File: plusOne.cpp
/**
 * Return an array of size *returnSize.
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* plusOne(int* digits, int digitsSize, int* returnSize) {
    if ( digits==NULL || digitsSize<1)
      return NULL;
    int i=digitsSize-1;
    int* newDigits=NULL;
    while( digits[i]+1 > 9 && i>=0 )
    {//从后往前扫描，直到遇到直到遇到某位数digits[i]+1<=9或者i<0时，停止循环
      digits[i]=0;
      i--;
    }
    if( i == -1 )  //溢出的情况
    {
      *returnSize=digitsSize+1;
      newDigits=(int*)malloc(sizeof(int)*(*returnSize));
      memcpy(newDigits+1,digits,digitsSize*sizeof(int));
      newDigits[0]=1;
    }
    else
    {
      *returnSize=digitsSize;
      newDigits=(int*)malloc(sizeof(int)*(*returnSize));
      memcpy(newDigits,digits,sizeof(int)*(*returnSize));
      newDigits[i]=digits[i]+1;
    }
    return newDigits;
}

File: plusOne.py
class Solution(object):
    def plusOne(self, digits):
        """
        :type digits: List[int]
        :rtype: List[int]
        """
        i=len(digits)-1
        while(digits[i]+1 > 9 and i >= 0): #从后往前扫描，
            digits[i]=0   #直到遇到某位数加1<=9停止
            i=i-1
        if i==-1 : #若循环结束之后，i==-1,说明digits的各位数字都为9
            digits.insert(0,1)#需要在第一位之前插入1即可
        else:#若i！=-1说明循环在中间的某位数上就停止了，该位数digits[i]+1<=9
            digits[i]=digits[i]+1 #只需要将该位数置为digits[i]=digits[i]+1
        return digits

li=[2,9]
s = Solution()
print s.plusOne(li)
